detector/etcd/client.go

```diff
@@ -0,0 +1,224 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package etcd
+
+import (
+	etcd "github.com/coreos/etcd/client"
+	log "github.com/golang/glog"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+// Etcd watch event actions
+const (
+	EtcdUpdate = "update"
+	EtcdSet    = "set"
+	EtcdCAS    = "compareAndSwap"
+)
+
+// Configurations for etcd client
+type ETCDClientCfg struct {
+	connectionTimeout time.Duration
+	username          string
+	password          string
+}
+
+type etcdClient struct {
+	client    etcd.KeysAPI
+	etcdError chan error
+	outgoing  chan *kvPair
+	ctx       context.Context
+	cancel    context.CancelFunc
+	stopped   bool
+	stopLock  sync.Mutex
+	done      chan struct{} // signal chan, closes when the underlying connection terminates
+}
+
+type kvPair struct {
+	key       string
+	value     []byte
+	lastIndex uint64
+}
+
+// Creates a new Etcd client with the specified list of endpoints and an optional configuration.
+func createEtcdClient(addrs []string, options *ETCDClientCfg) (*etcdClient, error) {
+	entries := createEndpoints(addrs, "http")
+	cfg := &etcd.Config{
+		Endpoints:               entries,
+		Transport:               etcd.DefaultTransport,
+		HeaderTimeoutPerRequest: 3 * time.Second,
+	}
+
+	// Set options
+	if options != nil {
+		if options.connectionTimeout != 0 {
+			setTimeout(cfg, options.connectionTimeout)
+		}
+		if options.username != "" {
+			setCredentials(cfg, options.username, options.password)
+		}
+	}
+
+	c, err := etcd.New(*cfg)
+	if err != nil {
+		log.Errorln(err.Error())
+		return nil, err
+	}
+
+	s := &etcdClient{
+		etcdError: make(chan error),
+		outgoing:  make(chan *kvPair),
+		client:    etcd.NewKeysAPI(c),
+		done:      make(chan struct{}),
+		stopped:   false,
+		ctx:       nil,
+		cancel:    nil,
+	}
+
+	go func() {
+		defer close(s.etcdError)
```

Comment:
this is racy if `Stop` is ever called while a watch is in progress. Possible that these chan's will close and then a write to either of them will panic